build.go

```diff
@@ -0,0 +1,110 @@
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/paketo-buildpacks/packit"
+	"github.com/paketo-buildpacks/packit/chronos"
+	"github.com/paketo-buildpacks/packit/postal"
+	"github.com/paketo-buildpacks/packit/scribe"
+)
+
+//go:generate faux --interface DependencyManager --output fakes/dependency_manager.go
+type DependencyManager interface {
+	Resolve(path, id, version, stack string) (postal.Dependency, error)
+	Deliver(dependency postal.Dependency, cnbPath, layerPath, platformPath string) error
+	GenerateBillOfMaterials(dependencies ...postal.Dependency) []packit.BOMEntry
+}
+
+//go:generate faux --interface NodeModuleBOM --output fakes/node_module_bom.go
+type NodeModuleBOM interface {
+	Generate(workingDir string) ([]packit.BOMEntry, error)
+}
+
+func Build(dependencyManager DependencyManager, nodeModuleBOM NodeModuleBOM, clock chronos.Clock, logger scribe.Emitter) packit.BuildFunc {
+	return func(context packit.BuildContext) (packit.BuildResult, error) {
+		logger.Title("%s %s", context.BuildpackInfo.Name, context.BuildpackInfo.Version)
+
+		logger.Process("Resolving CycloneDX Node.js Module version")
+
+		dependency, err := dependencyManager.Resolve(
+			filepath.Join(context.CNBPath, "buildpack.toml"),
+			"cyclonedx-node-module",
+			"*",
+			context.Stack,
+		)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+		logger.Subprocess("Selected %s version: %s", dependency.Name, dependency.Version)
+		logger.Break()
+
+		cycloneDXNodeModuleLayer, err := context.Layers.Get("cyclonedx-node-module")
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		cachedSHA, ok := cycloneDXNodeModuleLayer.Metadata["dependency-sha"].(string)
+		if ok && cachedSHA == dependency.SHA256 {
+			logger.Process("Reusing cached layer %s", cycloneDXNodeModuleLayer.Path)
+			logger.Break()
+		} else {
+			logger.Process("Executing build process")
+			cycloneDXNodeModuleLayer, err = cycloneDXNodeModuleLayer.Reset()
+			if err != nil {
+				return packit.BuildResult{}, err
+			}
+			logger.Subprocess("Installing %s %s", dependency.Name, dependency.Version)
+			duration, err := clock.Measure(func() error {
+				return dependencyManager.Deliver(dependency, context.CNBPath, cycloneDXNodeModuleLayer.Path, context.Platform.Path)
+			})
+			if err != nil {
+				return packit.BuildResult{}, err
+			}
+
+			logger.Action("Completed in %s", duration.Round(time.Millisecond))
+			logger.Break()
+
+			cycloneDXNodeModuleLayer.Metadata = map[string]interface{}{
+				"dependency-sha": dependency.SHA256,
+				"built_at":       clock.Now().Format(time.RFC3339Nano),
+			}
+		}
+
+		cycloneDXNodeModuleLayer.Cache = true
+
+		logger.Process("Configuring environment")
+		logger.Subprocess("Appending %s onto PATH", dependency.Name)
+		logger.Break()
+
+		os.Setenv("PATH", fmt.Sprint(os.Getenv("PATH"), string(os.PathListSeparator), filepath.Join(cycloneDXNodeModuleLayer.Path, "bin")))
+
+		toolBOM := dependencyManager.GenerateBillOfMaterials(dependency)
```

Comment:
Why is Tool-BOM and Module-BOM always generated? Can you not bypass them if `node_modules` haven't changed from a previous build by caching the tool+boms? That was my [question during the RFC review](https://github.com/paketo-buildpacks/rfcs/pull/98#discussion_r692172711).

I think we can expect users to keep re-building with simple application code changes and no module changes. if the node modules are large, we could save significant time if we can bypass this on rebuilds.

Reply:
To my understanding there is no way of caching the BOM at the current time, that is functionality that is being added by the upstream soon. We should rework how we do things once there is caching of BOM.

Reply:
My thought was this: Instead of generating the json file in the workspace, store it in a cached layer with the digest of node_modules as the cache key. If the node_modules haven't changed, just use it instead of downloading the tool & generating the bom.json. Does that not work? I'm also ok if you want to defer performance improvements for later. I wanted to clarify my comment on the RFC given that BOM generation seemed to add 3 seconds to an otherwise 5 sec build.

Reply:
Unless we also want to take the checksum of the node_modules directory in this buildpack as well. I don't know that there is a great way fo determining the modules haven't changed. I am willing to add this functionality but it seems that it might be easier to change fundamentally change the buildpack at that point to be split up into the different providers.

Reply:
I think we should support BOM re-use if we can, especially since we don't natively provide a way to opt out of the buildpack we should make it as efficient as we reasonable can

Reply:
Perhaps we should cut the release without this feature for now to unblock other BOM work, but add this as an issue to do next on this buildpack?

Reply:
Filed https://github.com/paketo-buildpacks/node-module-bom/issues/4 to address this down the road.